world.js

```diff
@@ -288,44 +25,140 @@ const _getBindSceneForRenderPriority = renderPriority => {
+      bindScene.add(app);
+
+      const hitTracker = hpManager.makeHitTracker();
+      hitTracker.bind(app);
+      app.dispatchEvent({type: 'hittrackeradded'});
+
+      const die = () => {
+        this.appManager.removeTrackedApp(app.instanceId);
+      };
+      app.addEventListener('die', die);
+    });
 
-  _bindHitTracker(app);
-});
-appManager.addEventListener('trackedappmigrate', async e => {
-  const {app, sourceAppManager, destinationAppManager} = e.data;
-  if (this === sourceAppManager) {
-    app.hitTracker.unbind();
-  } else if (this === destinationAppManager) {
-    _bindHitTracker(app);
+    // This handles removal of apps from the scene when we leave the world
+    this.appManager.addEventListener('appremove', async e => {
+      const app = e.data;
+      app.hitTracker.unbind();
+      app.parent.remove(app);
+    });
   }
-});
-appManager.addEventListener('appremove', async e => {
-  const app = e.data;
-  app.hitTracker.unbind();
-  app.parent.remove(app);
-});
\ No newline at end of file
+
+  isConnected() { return !!this.wsrtc; }
+
+  getConnection() { return this.wsrtc; }
+
+  // called by enterWorld() in universe.js
+  // This is called in single player mode instead of connectRoom
+  connectState(state) {
+    state.setResolvePriority(1);
+    playersManager.bindState(state.getArray(playersMapName));
+
+    this.appManager.unbindState();
+    this.appManager.clear();
+    const worldMap = state.getMap(worldMapName);
+    const appsArray = worldMap.get(appsMapName, Z.Array);
+
+    this.appManager.bindState(appsArray);
+
+    // Handy debug function to see the state
+    // Delete this eventually. For now press escape to see world state
+    document.addEventListener('keydown', event => {
+      if (event.key === 'Escape') {
+        console.log(state.getArray('world'));
+        console.log(state.getArray('players'));
+        console.log(scene);
+      }
+    });
+
+    const localPlayer = getLocalPlayer();
+    localPlayer.bindState(state.getArray(playersMapName));
+  }
+
+  // called by enterWorld() in universe.js
+  // This is called when a user joins a multiplayer room
+  // either from single player or directly from a link
+  async connectRoom(u, state = new Z.Doc()) {
+    // Players cannot be initialized until the physx worker is loaded
+    // Otherwise you will receive allocation errors because the module instance is undefined
+    await physx.waitForLoad();
+    await physxWorkerManager.waitForLoad();
+    const localPlayer = getLocalPlayer();
+
+    state.setResolvePriority(1);
+
+    // Create a new instance of the websocket rtc client
+    // This comes from webaverse/wsrtc/wsrtc.js
+    this.wsrtc = new WSRTC(u, {
+      localPlayer,
+      crdtState: state,
+    });
+
+    // Handy debug function to see the state
+    // Delete this eventually. For now press escape to see world state
+    document.addEventListener('keydown', event => {
```

Comment:
We should centralize these hacks.

There is already a place for it: https://github.com/webaverse/app/blob/6ab3bb7d2da38454e71c92105e6c7aad4b6ea13c/webaverse.js#L545